hibike/hibike_simulator.py

```diff
@@ -1,133 +1,156 @@
+            uid = values[0]
+            if uid in uid_to_queue:
+                del uid_to_queue[uid]
+            uid_to_queue[uid] = instruction_queue
+            res = ["device_subscribed", values]
         else:
-            return low + (random.random() - .5) * noise
-
-############
-## MOTORS ##
-############
-
-    def readValue(self, UID, param):
-        #param values: 0 - delay, 1 - last_time, 2 - data
+            print("Some error code received")
+            print(result, values)
+
+        if res is not None:
+            stateQueue.put(res)
+
+def runFakeSubscription(uid, fake_device_queue, fake_device_readings):
+    # fake_device_queue is input commands
+    # fake_device_readings is output of fake readings
+    timeout = None
+    new_delay = False
+    while True:
         try:
-            return self.subscribedTo[UID][param]
-        except:
-            return 1
-
-    def writeValue(self, UID, param, value):
-        if param >= len(self.subscribedTo[UID]):
-            return 1
-        self.subscribedTo[UID][param] = value
-        return 0
+            delay = fake_device_queue.get(block=True, timeout=timeout)
+            new_delay = True
+        except queue.Empty:
+            pass
+        # Simulate enumeration
+        if delay == 0:
+            params = list(paramMap[uid_to_device_id(uid)].keys())
+            fake_device_readings.put(("device_subscribed", [uid, 0, params]))
+        # Simulate subscription
+        elif delay > 0:
+            if new_delay:
+                params = fake_device_queue.get()
+            param_types = [paramMap[uid_to_device_id(uid)][param][1] for param in params]
+            params_and_values = {}
+            if __name__ == "__main__":
+                for param, param_type in zip(params, param_types):
+                    if param_type in ("bool", ):
+                        params_and_values[param] = random.choice([True, False])
+                    elif param_type in ("uint8_t", "int8_t", "uint16_t", "int16_t", "uint32_t", "int32_t", "uint64_t", "int64_t"):
+                        params_and_values[param] = random.randrange(256)
+                    else:
+                        params_and_values[param] = random.random()
+            else:
+                for param, param_type in zip(params, param_types):
+                    if param_type in ("bool", ):
+                        params_and_values[param] = True
+                    elif param_type in ("uint8_t", "int8_t", "uint16_t", "int16_t", "uint32_t", "int32_t", "uint64_t", "int64_t"):
+                        params_and_values[param] = 123
+                    else:
+                        params_and_values[param] = 101
+            fake_device_readings.put(("device_values", [uid, list(params_and_values.items())]))
+            timeout = delay / 1000.0
+            new_delay = False
+        else:
+            print("Invalid negative delay: {:d}".format(delay))
+
+class FakeDevice():
```

Comment:
I would find that the file reads better if the code for `FakeDevice` were moved above the `run*` functions. What do you think?